src/main/java/com/google/sps/data/ListHelper.java

```diff
@@ -0,0 +1,93 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.data;
+
+import com.google.gson.Gson;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Collections;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.datastore.QueryResultList;
+import com.google.appengine.api.datastore.Cursor;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.Query.Filter;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import com.google.appengine.api.datastore.Query.CompositeFilter;
+import com.google.appengine.api.datastore.Query.CompositeFilterOperator;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import javax.servlet.http.HttpServletRequest;
+import com.google.sps.data.GivrUser;
+import java.io.IOException;
+
+public abstract class ListHelper {
```

Comment:
Argh... github keeps losing my comments!

While this works, it's not quite what I was thinking. The class here only has static functions; you don't need to mark the class as abstract in that case (the above specializations can just call the static functions).

For posterity, here's what I was thinking:

```

public abstract class ListHelper {
  protected String entityKind;
  protected HttpServletRequest request;
  protected GivrUser currentUser;

  ListHelper(String entityKind, HttpServletRequest request, GivrUser currentUser) {
    // store all args...
  }

  public Query getQuery() {
    return this.getQueryFromFilters(
        this.handleUserFiltering(coerceParameterToBoolean(this.request, "displayForUser")));
  }

  // Impemented by the subclass.
  // Gets the constant map that translates from known key -> datastore property for the entity.
  abstract HashMap<String, String> GetDatastoreConstantMap();
  // Returns the filter for this.entity against this.user
  abstract ArrayList<Filter> handleUserFiltering(boolean displayForUser);

  // Common functions used for query.
  private static HashMap<String, ArrayList<String>> parseFilterParams() {
    // old code with the following changes:
    // datastore map = this.GetDatastoreConstantMap();
    // you can get the request from: this.request
  }

  /* This function constructs a query based on the entity kind, request parameters and any previously added or new filters */
  private static Query getQueryFromFilters(
      String entityKind,
      HashMap<String, ArrayList<String>> filterParamMap,
      ArrayList<Filter> filterCollection)
      throws IllegalArgumentException {
    // same as what you have, except you don't pass in args.
    // filterParamMap = this.parseFilterParams();
    // everything else is in instance vars.
  }

  public static boolean coerceParameterToBoolean(HttpServletRequest request, String key) {
    String requestParameter = request.getParameter(key);
    return (requestParameter != null) && (requestParameter.equals("true"));
  }
}

public class ListEventsHelper extends ListHelper {
  HashMap<String, String> GetDatastoreConstantMap() {
    // return a new instance of the constnat map here instead of in the servlet.
  }

  public ArrayList<Filter> handleUserFiltering(boolean displayForUser) {
    // basically your old functions, but use the instance vars.
  }
}

public class ListOrganizationsHelper extends ListHelper {
  HashMap<String, String> GetDatastoreConstantMap() {
    // return a new instance of the constnat map here instead of in the servlet.
  }

  public ArrayList<Filter> handleUserFiltering(boolean displayForUser) {
    // basically your old functions, but use the instance vars.
  }
}


```